WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift

```diff
@@ -0,0 +1,146 @@
+    }()
+
+    lazy private(set) var previousRangeDescription: String = {
+        generateDescriptionOf(timeRange: previousTimeRange)
+    }()
+
+    var selectionDescription: String {
+        selectionType.description
+    }
+
+    init(selectedTimeRange: StatsTimeRangeV4,
+         currentDate: Date = Date(),
+         currentCalendar: Calendar = Calendar(identifier: .iso8601)
+    ) {
+        self.selectionType = SelectionType.from(selectedTimeRange)
+        self.currentDate = currentDate
+        self.currentCalendar = currentCalendar
+    }
+
+    private func generateCurrentTimeRangeFrom(selectionType: SelectionType) -> TimeRange {
+        switch selectionType {
+        case .today:
+            return TimeRange(start: currentDate.startOfDay(timezone: currentTimezone), end: currentDate)
+        case .weekToDate:
+            let weekStart = currentDate.startOfWeek(timezone: currentTimezone, calendar: currentCalendar)
+            return TimeRange(start: weekStart, end: currentDate)
+        case .monthToDate:
+            return TimeRange(start: currentDate.startOfMonth(timezone: currentTimezone), end: currentDate)
+        case .yearToDate:
+            return TimeRange(start: currentDate.startOfYear(timezone: currentTimezone), end: currentDate)
+        }
+    }
+
+    private func generatePreviousTimeRangeFrom(selectionType: SelectionType) -> TimeRange {
+        switch selectionType {
+        case .today:
+            let oneDayAgo = currentCalendar.date(byAdding: .day, value: -1, to: currentDate)!
+            return TimeRange(start: oneDayAgo.startOfDay(timezone: currentTimezone), end: oneDayAgo)
+        case .weekToDate:
+            let oneWeekAgo = currentCalendar.date(byAdding: .day, value: -7, to: currentDate)!
+            let lastWeekStart = oneWeekAgo.startOfWeek(timezone: currentTimezone, calendar: currentCalendar)
+            return TimeRange(start: lastWeekStart, end: oneWeekAgo)
+        case .monthToDate:
+            let oneMonthAgo = currentCalendar.date(byAdding: .month, value: -1, to: currentDate)!
+            return TimeRange(start: oneMonthAgo.startOfMonth(timezone: currentTimezone), end: oneMonthAgo)
+        case .yearToDate:
+            let oneYearAgo = currentCalendar.date(byAdding: .year, value: -1, to: currentDate)!
+            return TimeRange(start: oneYearAgo.startOfYear(timezone: currentTimezone), end: oneYearAgo)
+        }
+    }
+
+    private func generateDescriptionOf(timeRange: TimeRange) -> String {
+        let dateFormatter = DateFormatter()
+
+        if selectionType == .today {
+            dateFormatter.dateFormat = "MMM d, yyyy"
+            return dateFormatter.string(from: timeRange.start)
+        }
+
+        dateFormatter.dateFormat = "MMM d"
+        let startDateDescription = dateFormatter.string(from: timeRange.start)
+        let endDateDescription = generateEndDateDescription(timeRange: timeRange, dateFormatter: dateFormatter)
```

Comment:
We aim to have all our date formats in the `DateFormatters + Woo` extension. This is to make sure we are somehow consistent on our date formats. Can we reuse some of those? If not, it's probably good to add these there!

Reply:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/861257e4131b5bac244687b0340f64abe8135437